### PR TITLE
vo_opengl: enable compute shader for mesa

### PR DESCRIPTION
**Pull request by bjin**
Mesa 17.1 supports compute shader but not full specs of OpenGL 4.3.
Change the code to detect OpenGL extension `GL_ARB_compute_shader`
rather than OpenGL version `4.3`.

The compute polar code also utilized OpenGL extension
`GL_ARB_arrays_of_arrays`, so add it as a requirement as well.

Tested with Intel HD 4000 graphics card on Linux (Mesa 17.1.4). Upscaling
1280x720 video to 1920x1080 with `scale=ewa_lanczossharp`.

    pass rendering time: 46.3 ms -> 18.5ms

### CODE REVIEW

**Comment by haasn** [comment]
Instead of `GL_ARB_array_of_arrays` we could also try using a single dimensional array and indexing it at `[stride * y + x]` instead of `[y][x]`. Can you test if that makes a performance difference? Because if so, I would prefer that approach to depending on this extension.

File: video/out/opengl/video.c [hunk 1]
```diff
@@ -1183,7 +1183,7 @@ static void dispatch_compute(struct gl_video *p, int w, int h, int bw, int bh)
         // Clamp the texture coordinates to prevent sampling out-of-bounds in
         // threads that exceed the requested width/height
         PRELUDE("#define texmap%d(id) min(texcoord%d_rot(id), vec2(1.0))\n", n, n);
-        PRELUDE("const vec2 texcoord%d = texmap%d(gl_GlobalInvocationID);\n", n, n);
+        PRELUDE("#define texcoord%d texmap%d(gl_GlobalInvocationID)\n", n, n);
```
Comment (haasn) [inline video/out/opengl/video.c]: Why this change? Seems like it would make the texcoord be recomputed multiple times, instead of reused once.

Comment (bjin) [inline video/out/opengl/video.c]: mesa complains:

     error: initializer of const variable `texcoord0' must be a constant expression

Comment (haasn) [inline video/out/opengl/video.c]: Hmm, seems like making it a `const` breaks e.g. AMD as well. Could we also just drop the `const` keyword?

**Comment by haasn** [comment]
Btw, it's possible that our check for `hdr_compute_peak` can be similarly relaxed. I just felt like all of the platforms that implement compute shaders and SSBOs will also implement OpenGL 4.3, so I didn't give it much though.

File: video/out/opengl/video.c (hunk 1 above)
Comment (bjin) [inline video/out/opengl/video.c]: It works. I will just drop `const` then.

**Comment by haasn** [comment]
Btw, I would love to try and see if compute shaders can make regular scaling (orthogonal) faster too. The potential gains could be huge, depending on how much we can benefit from parallelism when doing long/thin work groups.

Since EWA is still impractical on iGPUs, it might be a better idea to focus on that.

**Comment by bjin** [comment]
> Can you test if that makes a performance difference? Because if so, I would prefer that approach to depending on this extension.

No performance drop. Do you want me change it to regular array in this PR?

> Btw, it's possible that our check for hdr_compute_peak can be similarly relaxed. 

I thought about this, but I don't have hdr content to test it (mesa supports SSBO as well).

**Comment by Gusar321** [comment]
> Btw, it's possible that our check for hdr_compute_peak can be similarly relaxed. I just felt like all of the platforms that implement compute shaders and SSBOs will also implement OpenGL 4.3, so I didn't give it much though.

I think the same applies for direct rendering. It checks for OpenGL 4.4, but there are drivers which don't have that, yet they do have GL_ARB_buffer_storage

**Comment by haasn** [comment]
> Do you want me to change it to regular array in this PR?

Please do.

> I thought about this, but I don't have hdr content to test it (mesa supports SSBO as well).

Just play any clip with `--vf=format=gamma=hlg` and it'll magically be HDR. (Also add `--hdr-compute-peak` to test that code. How well does it perform?)

**Comment by haasn** [comment]
@Gusar321 If you can identify hardware where 1) it would work otherwise and 2) results in a performance increase, feel free to open a PR for that check as well.

**Comment by bjin** [comment]
Pushed new commits. Please take a look (especially for the `stride` change).

>  Just play any clip with --vf=format=gamma=hlg and it'll magically be HDR. (Also add --hdr-compute-peakto test that code. How well does it perform?)

It seems to work with SSBO enabled (no error at least), `10ms` with the same intel card.

**Comment by haasn** [comment]
Amusingly enough it does actually seem to be faster using the `[][]` arrays compared to the 1D array, but the difference is extremely insignificant (3740 vs 3720 for radius=4)

File: video/out/opengl/video.c [hunk 2]
```diff
@@ -2464,6 +2464,8 @@ static void pass_colormanage(struct gl_video *p, struct mp_colorspace src, bool
             gl->BindBuffer(GL_SHADER_STORAGE_BUFFER, 0);
         }
 
+        gl_sc_enable_extension(p->sc, "GL_ARB_shader_storage_buffer_object");
+
```
Comment (haasn) [inline video/out/opengl/video.c]: Cosmetic: Probably better without the newline. Also should we make gl_sc_ssbo implicitly turn on the extension instead of relying on the user to specify it?

Comment (bjin) [inline video/out/opengl/video.c]: Yes, make sense to me (to move it inside of `gl_sc_ssbo`)

**Comment by bjin** [comment]
> Amusingly enough it does actually seem to be faster using the [][] arrays compared to the 1D array, but the difference is extremely insignificant (3740 vs 3720 for radius=4)

So, bring it back?

File: video/out/opengl/video.c [hunk 3]
```diff
@@ -3391,7 +3393,8 @@ static void check_gl_features(struct gl_video *p)
     bool have_mglsl = gl->glsl_version >= 130; // modern GLSL (1st class arrays etc.)
     bool have_texrg = gl->mpgl_caps & MPGL_CAP_TEX_RG;
     bool have_tex16 = !gl->es || (gl->mpgl_caps & MPGL_CAP_EXT16);
-    bool have_compute = gl->glsl_version >= 430; // easiest way to ensure all
+    bool have_compute = gl->DispatchCompute;
```
Comment (haasn) [inline video/out/opengl/video.c]: I think the problem with doing the check for just `gl->DispatchCompute` is that we also need `GL_ARB_shader_image_load_store`. The same goes for the check up above (for EWA).

I don't think `GL_ARB_compute_shader` implies `GL_ARB_shader_image_load_store`.

Comment (bjin) [inline video/out/opengl/video.c]: Then check both `gl->DispatchCompute` and `gl->BindImageTexture` ?

Comment (haasn) [inline video/out/opengl/video.c]: Maybe it would be better to cache these capabilities in the `gl->caps` field, though. Basically introducing `MPGL_CAP_COMPUTE` which would be defined as both `GL_ARB_shader_image_load_store` and `GL_ARB_compute_shader` and whatever else is necessary (e.g. 2D arrays), since I think we will eventually need 2D arrays anyway. Then we can drop the more confusing/slower `stride` code again.

**Comment by haasn** [comment]
> It seems to work with SSBO enabled (no error at least), 10ms with the same intel card.

Damn, that's slower than I'd hope for it to be. It seems like the nvidia compiler does a really good job of optimizing away atomics though, which is probably why it's so fast. (Even just hitting a single global atomic is like a few 100 μs on nvidia, compared to 10-20ms when unoptimized (e.g. inside a fragment shader))

I wish I knew how to write code as fast as the nvidia implementation so we could use it for all platforms, but everything I tried ended up slower (probably because of missing warp exchange primitives).

File: video/out/opengl/video.c (hunk 3 above)
Comment (bjin) [inline video/out/opengl/video.c]: > Basically introducing MPGL_CAP_COMPUTE which would be defined as both GL_ARB_shader_image_load_store and GL_ARB_compute_shader

It seems to require new routines in `mpgl_load_functions2()` dedicated for this purpose, instead of simple definition in `gl_functions` list. Not sure if it's worth doing so.

Comment (haasn) [inline video/out/opengl/video.c]: I don't think it would be a big deal, only a one line definition that updates `gl->mpgl_caps`. Seems to fit right next to this one:

```glsl
    if (is_software_gl(gl)) {
        gl->mpgl_caps |= MPGL_CAP_SW;
        mp_verbose(log, "Detected suspected software renderer.\n");
    }
```

I guess it depends on how often we're going to be accessing these things. I would certainly love to try and add more compute-shader optimized drop-in-replacement routines in the future, so any additional check that we can reduce to a single flag comparison would benefit from it. And since we're already checking for three extensions in total (compute_shader, image_load_store and array_of_arrays) that's worth it, I think.

Comment (bjin) [inline video/out/opengl/video.c]: The interesting part is that `GL_ARB_arrays_of_arrays` is a GLSL only extensions, there is no function to be loaded (no `gl->DispatchCompute`-like check). I also believe `GL_ARB_arrays_of_arrays` should be a separated capability, it's not a strict subset of compute shader I would say.

Comment (haasn) [inline video/out/opengl/video.c]: I guess, but I think most uses of compute shaders are going to end up using 2D arrays, especially all of the user shaders. In fact it pretty much seems to only be the HDR detection code so far that doesn't need them. I'd prefer simpler checks that are more readable and more maintainable over trying to get the exact minimum condition right. If even `mesa` supports array_of_array then I can't imagine it would make much of a difference.

Comment (haasn) [inline video/out/opengl/video.c]: Actually another way you could do it is like this:

```glsl
    MPGL_CAP_SSBO               = (1 << 21),    // GL_ARB_shader_storage_buffer_object
    MPGL_CAP_COMPUTE            = (1 << 22),    // GL_ARB_compute_shader
    MPGL_CAP_2D_ARRAY           = (1 << 23),    // GL_ARB_array_of_arrays
```

And then we could have something like “#define CAP_COMPUTE (MPGL_CAP_COMPURE | MPGL_CAP_2D_ARRAY)` inside video.c which we can use to simplify our repeated checks for multiple features? Hard to call.

Comment (bjin) [inline video/out/opengl/video.c]: Actually, my point is, even if we want to combine these three extension as a whole, there is no easy way. With no function provided by `GL_ARB_arrays_of_arrays`, it's just simpler and easier to use a new capability for 2D array.

**Comment by bjin** [comment]
Pushed new change, with two new capability for both compute shader and 2D array.

**Comment by haasn** [comment]
Alright, looks fine to me.

**Comment by haasn** [comment]
Merged, thanks. That should fix the new EWA code on AMD etc. as well.

**Comment by haasn** [comment]
@bjin
Seems like this EWA compute shader can break on mesa+radeonsi, potentially some other systems. Maybe there's some other dependency that we need? Or those systems are just broken?

**Comment by haasn** [comment]
Actually I think I may have spotted the issue: image accesses from compute shaders are non-coherent, but we never insert a memory barrier in between the compute shader and the following shaders that sample from it again. So wrong data could still be cached, or perhaps the image layout would be incorrect. Since AMD supports async compute that might explain why I don't get the bug locally. Nvidia also doesn't seem to give a shit about coherency or texture layouts in general, it's all just the same for them.

**Comment by haasn** [comment]
Hmm nope issue persists even with the correct memory barrier inserted
